tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+            ASSERT_EQ(msg[i], pt[i]) << "Message and plaintext do not match";
+    }
+
+    TEST_F(CompletenessTest, AsymEnc) {
+        // Encrypt the message
+        std::vector<uint8_t> ct(crypto.AsymEncSize(msg.size()));
+        ASSERT_EQ(0,
+                  crypto.AsymEnc(pem_key, msg.data(), ct.data(), msg.size()));
+
+        // Decrypt the message
+        std::vector<uint8_t> pt(crypto.AsymDecSize(ct.size()));
+        size_t _pt_size;
+        ASSERT_EQ(0,
+                  crypto.AsymDec(ct.data(), pt.data(), ct.size(), &_pt_size));
+
+        for (int i = 0; i < msg.size(); i++)
+            ASSERT_EQ(msg[i], pt[i]) << "Message and plaintext do not match";
+    }
+
+    TEST_F(CompletenessTest, Sig) {
+        // Sign the message
+        std::vector<uint8_t> sig(crypto.AsymSignSize());
+        ASSERT_EQ(0, crypto.Sign(msg.data(), sig.data(), msg.size()));
+
+        // Verify the signature
+        ASSERT_EQ(0, crypto.Verify(pem_key, msg.data(), sig.data(), msg.size()))
+            << "Signature failed to verify";
+    }
+
+#ifdef HOST
+    TEST_F(CompletenessTest, SigUsingKeyfile) {
+        // Sign the message
+        std::vector<uint8_t> sig(crypto.AsymSignSize());
+        ASSERT_EQ(0,
+                  crypto.SignUsingKeyfile("../../tests/data/private.pem",
+                                          msg.data(), sig.data(), msg.size()));
+
+        // Load public verification key from file
+        std::vector<uint8_t> pub_key;
+        load_file(std::string("../../tests/data/public.pub"), pub_key);
+
+        // Verify the signature
+        ASSERT_EQ(0, crypto.Verify(pub_key.data(), msg.data(), sig.data(),
+                                   msg.size()));
+    }
+#endif
+
+    TEST_F(CompletenessTest, Attestation) {
+        // Currently the unittests only support building for the host, so we
+        // can't generate evidence. Instead we verify pre-generated evidence
+
+        // Load nonce, public key, enclave signing key, and evidence from file
+        std::vector<uint8_t> nonce;
+        std::vector<uint8_t> pub_key;
+        std::vector<uint8_t> signing_key;
+        std::vector<uint8_t> evidence;
+        load_file(std::string("../../tests/data/nonce"), nonce);
+        load_file(std::string("../../tests/data/report_key.pub"), pub_key);
+        load_file(std::string("../../tests/data/signing_key.pub"), signing_key);
+        load_file(std::string("../../tests/data/evidence"), evidence);
+
+        // Verify the evidence
+        ASSERT_EQ(0, attestation->AttestEvidence(
+                         &sgx_remote_uuid, signing_key.data(), evidence.data(),
+                         pub_key.data(), nonce.data(), signing_key.size() + 1,
+                         evidence.size(), pub_key.size()));
+    }
+
+    // ###########################
+    // ##### SOUNDNESS TESTS #####
+    // ###########################
+
+    // Dummy class to create a new test suite
+    class SoundnessTest : public UtilsTest {};
```

Comment:
not sure how useful this would be, but maybe for each test we can also mess with the size parameters for each function, e.g. passing in `msg.size() + 1` when we should be passing in `msg.size()`

Reply:
For which tests/parameters exactly? I'm not sure I totally see the use as this would either segfault, or just lead to an incorrect ending result? Or maybe I'm misunderstanding

Reply:
oh yeah you're probably right about the segfault. i had thought that passing in an incorrect size for the buffer to encrypt would get the `SymEnc()` (for example) function to return an error, but it would probably just segfault. never mind!